### PR TITLE
[useSES/extra] Reuse old selection if possible

### PR DESCRIPTION
**Pull request by acdlite**
When you pass an unmemoized selector to useSyncExternalStoreExtra, we have to reevaluate it on every render because we don't know whether it depends on new props.

However, after reevalutating it, we should still compare the result to the previous selection with `isEqual` and reuse the old one if it hasn't changed.

Originally I did not implement this, because if the selector changes due to new props or state, the component is going to have to re-render anyway. However, it's still useful to return a memoized selection when possible, because it may be the input to a downstream memoization.

In the test I wrote, the example I chose is selecting a list of items from the store, and passing the list as a prop to a memoized component. If the list prop is memoized, we can bail out.

### CODE REVIEW

**Review by rickhanlonii — APPROVED** [review]

File: packages/use-sync-external-store/src/__tests__/useSyncExternalStoreShared-test.js [hunk 1]
```diff
@@ -666,4 +670,81 @@ describe('Shared useSyncExternalStore behavior (shim and built-in)', () => {
...
+        </ul>
+      );
+    });
+
+    function App({step}) {
+      const inlineSelector = state => {
+        Scheduler.unstable_yieldValue('Inline selector');
+        return [...state.items, 'C'];
+      };
+      const items = useSyncExternalStoreExtra(
```
Comment (gaearon) [inline packages/use-sync-external-store/src/__tests__/useSyncExternalStoreShared-test.js:708]: ah, connectAdvanced

**Comment by markerikson** [comment]
FYI, the shim can't actually use the built-in version atm, because it's not looking for the `unstable_` prefix:

`builtInAPI = React.useSyncExternalStore`

**Comment by rickhanlonii** [comment]
Just following up: that issue was fixed, and we're checking for unstable now.

**Comment by zeorin** [comment]
@acdlite why did you assign `instRef.current` to `inst`?

Seems they both keep pointing to the same object throughout the rest of the code in this hook, but the `react-hooks/exhaustive-deps` lint rule (correctly) complains about `inst` not being in the dependency array.

It seems to me like one could do away with the reassignment and just access the object using `instRef.current` directly, which would satisfy the linter; and also just generally be more idiomatic and easier to understand.

Or am I missing something?

**Pull request merged**